git-tutorials/git.md

```diff
@@ -0,0 +1,78 @@
+
+For example, if we wanted to change to our Documents directory, we would do the following:
+`cd Documents`
+
+To clone our repository, we use the following command in our terminal:
+`git clone <https address here>`
+
+For example:
+`git clone https://github.com/your-username/open.git fiveable-open`
+
+Again, this will create a local copy of the repository at the directory specified earlier. You will then need to open these files in your preferred text editor to make your changes.
+
+### Checkout
+It is highly reccomended to switch to a new branch to make changes instead of committing to the master branch. Why? Say you commit to the master branch of your fork. You go ahead and make a PR, but oh no! Your PR was denied and was closed. You now have a problem, your master branch now has a bunch of commits that don't exist on the upstream repo. **Your fork is now out of sync with the upstream repo.** You could just revert those commits, but now you have a bunch of revert commits polluting your commit history, and those will be added to the upstream repo if you make another PR. Or you could rebase and remove history (which would a require a force push which we do NOT allow on this repo). Both of these options are not very good and will only cause hassle to the maintainers and to the contributor themselves.
+
+Sounds like a lot right? It's better to prevent all of this from happening in the beginning by just making changes to a new branch every time. So say I want to add a bunch of badges to the readme, I'd make a branch `chore/readme-branch` or `chore/patch-1.` 
+
+To create a new branch from your terminal, do the following:
+* `git branch <branchname>` to create a new branch
+* `git checkout <branchname>` to swtich to your branch
+
+If you make a bunch of commits and forget to switch branches, you can use `git switch <branchname>` or `git switch -c <branchname>` which will switch to another branch and move your already-done commits to this new branch. Appending `-c` will create a new branch while `git switch <branchname>` switches to an existing branch.
+ 
+### Pull
+You will need to use `git pull` to fetch the latest changes from another repository or local branch. When you run `git pull` in your terminal, remote content will be downloaded and immediately change the local state of your repository to match the new content. It is strongly recommended to not `git pull` when you have uncommitted changes. 
+
+In the event that you do have uncommitted changes and want to `git pull`, you will need to do the following in your command line. First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash apply` to apply the changes back onto your local copy, and then `git stash drop` to remove those changes from the stash stack. This method is not reccommened as unexpected behavior may occur.
```

Comment:
```suggestion
In the event that you do have uncommitted changes and want to `git pull`, you will need to do the following in your command line. First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash apply` to apply the changes back onto your local copy, and then `git stash drop` to remove those changes from the stash stack. 
```
💀 `git stash pop` is what is not recommended and what'll cause unexpected behavior, `git stash apply` & `git stash drop` are the alternative.